x/config/chooser.go

```diff
@@ -0,0 +1,204 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package config
+
+import (
+	"fmt"
+
+	"go.uber.org/yarpc/api/peer"
+	"go.uber.org/yarpc/internal/decode"
+	peerbind "go.uber.org/yarpc/peer"
+	"go.uber.org/yarpc/peer/x/peerheap"
+	"go.uber.org/yarpc/peer/x/roundrobin"
+)
+
+// ChooserConfig facilitates decoding and building peer choosers.
+// The peer chooser combines a peer list (for the peer selection strategy, like
+// least-pending or round-robin) with a peer list binder (like static peers or
+// dynamic peers from DNS or watching a file in a particular format).
+type ChooserConfig struct {
+	With   string   `config:"with"`   // like peers / file / dns / dns-srv
+	Choose string   `config:"choose"` // like round-robin / least-pending
+	Peer   string   `config:"peer"`   // implies with: peer
+	Peers  []string `config:"peers"`  // implies with: peers
+	Etc    attributeMap
+}
+
+// Decode captures a configuration on this ChooserConfig struct.
+func (c ChooserConfig) Decode(into decode.Into) error {
+	var err error
+
+	c.With, err = c.Etc.PopString("with")
+	if err != nil {
+		return fmt.Errorf(`could not decode outbound peer list binder config, "with": %v`, err)
+	}
+
+	c.Choose, err = c.Etc.PopString("choose")
```

Comment:
do the errors here include the string not being present?  From what it looks like this decoder requires every field to be present or it returns an error >.>